### PR TITLE
add random to transcript generated ids and set sent activity.id

### PR DESCRIPTION
**Pull request by EricDahlvang**
Fixes https://github.com/microsoft/botbuilder-js/issues/1340

Also add random with prefix to log generated ids.

### CODE REVIEW

File: libraries/botbuilder-core/src/turnContext.ts [hunk 1]
```diff
@@ -429,6 +429,13 @@ export class TurnContext {
                 .then((responses: ResourceResponse[]) => {
                     // Set responded flag
                     if (sentNonTraceActivity) { this.responded = true; }
+                    if(responses) {
+                        output.map((a: Partial<Activity>, index: number) => {
+                            if (index < responses.length) {
+                                a.id = responses[index].id;
+                            }
+                        });
+                    }
```
Comment (stevengum) [inline libraries/botbuilder-core/src/turnContext.ts]: This all looks good, but can you add a test for this?

File: libraries/botbuilder-core/src/turnContext.ts [hunk 2]
```diff
@@ -420,7 +420,7 @@ export class TurnContext {
             const o: Partial<Activity> = TurnContext.applyConversationReference({...a}, ref);
             if (!o.type) { o.type = ActivityTypes.Message; }
             if (o.type !== ActivityTypes.Trace) { sentNonTraceActivity = true; }
-
+            if (o.id) { delete o.id; }
```
Comment (stevengum) [inline libraries/botbuilder-core/src/turnContext.ts]: Can you add a test for this? The bot sends the same activity twice and the id should be different on both activities.

File: libraries/botbuilder-core/src/transcriptLogger.ts [hunk 3]
```diff
@@ -52,21 +52,19 @@ export class TranscriptLoggerMiddleware implements Middleware {
...
+                // For certain channels, a ResourceResponse with an id is not always sent to the bot.
+                // This fix uses the timestamp on the activity to populate its id for logging the transcript.
+                // If there is no outgoing timestamp, the current time for the bot is used for the activity.id.
+                // See https://github.com/microsoft/botbuilder-js/issues/1122
+                if (!clonedActivity.id) {
+                    const prefix = `g_${Math.random().toString(36).slice(2,8)}`;
+                    if (clonedActivity.timestamp) {
+                        clonedActivity.id = `${prefix}${new Date(clonedActivity.timestamp).getTime().toString()}`;
+                    } else {
+                        clonedActivity.id = `${prefix}${Date.now().toString()}`;
```
Comment (Stevenic) [inline libraries/botbuilder-core/src/transcriptLogger.ts]: did you mean to not call getTime() here?

File: libraries/botbuilder-core/src/turnContext.ts [hunk 4]
```diff
@@ -429,6 +429,13 @@ export class TurnContext {
                 .then((responses: ResourceResponse[]) => {
                     // Set responded flag
                     if (sentNonTraceActivity) { this.responded = true; }
+                    if(responses) {
+                        output.map((a: Partial<Activity>, index: number) => {
+                            if (index < responses.length) {
+                                a.id = responses[index].id;
+                            }
+                        });
+                    }
```
Comment (Stevenic) [inline libraries/botbuilder-core/src/turnContext.ts]: what's the point of this? The output has been sent and is about to be thrown away...

Comment (EricDahlvang) [reply libraries/botbuilder-core/src/turnContext.ts]: Yeah, this is actually supposed to be:
```js
                        activities.map((a: Partial<Activity>, index: number) => {
                            if (index < responses.length) {
                                a.id = responses[index].id;
                            }
                        });
```

Comment (EricDahlvang) [reply libraries/botbuilder-core/src/turnContext.ts]: removed

File: libraries/botbuilder-core/src/transcriptLogger.ts [hunk 5]
```diff
@@ -52,23 +52,23 @@ export class TranscriptLoggerMiddleware implements Middleware {
...
-                        // If there is no outgoing timestamp, the current time for the bot is used for the activity.id.
-                        // See https://github.com/microsoft/botbuilder-js/issues/1122
-                        if (!clonedActivity.id) {
-                            if (clonedActivity.timestamp) {
-                                clonedActivity.id = new Date(clonedActivity.timestamp).getTime().toString();
-                            } else {
-                                clonedActivity.id = Date.now().toString();
-                            }
-                        }
+                if (index < responses.length && !clonedActivity.id) {
```
Comment (stevengum) [inline libraries/botbuilder-core/src/transcriptLogger.ts]: We should always set the resourceResponse Id on the activity to be logged down here.

**Comment by stevengum** [comment]
I've removed the label for 4.6 for now since it doesn't seem like we're going to add this to our next RCs. @cleemullins, @sgellock if that's not kosher please relabel this PR.

**Comment by sgellock** [comment]
@stevengum sounds right.  Thank you

File: libraries/botbuilder-core/src/transcriptLogger.ts (hunk 5 above)
Comment (EricDahlvang) [reply libraries/botbuilder-core/src/transcriptLogger.ts]: Good point ... removed '!clonedActivity.id'

**Pull request merged**